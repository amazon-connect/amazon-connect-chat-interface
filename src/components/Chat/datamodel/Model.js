// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

export const PARTICIPANT_MESSAGE = "MESSAGE";
export const CHAT_EVENT = "EVENT";
export const ATTACHMENT_MESSAGE = "ATTACHMENT";

export const PARTICIPANT_TYPES = {
    AGENT: "AGENT",
    CUSTOMER: "CUSTOMER"
};

export const ContentType = {
    EVENT_CONTENT_TYPE: {
        TYPING: "application/vnd.amazonaws.connect.event.typing",
        PARTICIPANT_JOINED: "application/vnd.amazonaws.connect.event.participant.joined",
        PARTICIPANT_LEFT: "application/vnd.amazonaws.connect.event.participant.left",
        TRANSFER_SUCCEEDED: "application/vnd.amazonaws.connect.event.transfer.succeeded",
        TRANSFER_FAILED: "application/vnd.amazonaws.connect.event.transfer.failed",
        CONNECTION_ACKNOWLEDGED: "application/vnd.amazonaws.connect.event.connection.acknowledged",
        CHAT_ENDED: "application/vnd.amazonaws.connect.event.chat.ended"
    },
    MESSAGE_CONTENT_TYPE: {
        TEXT_PLAIN: "text/plain",
        APPLICATION_PDF: "application/pdf",
        IMAGE_JPG: "image/jpeg",
        IMAGE_PNG: "image/png",
        APPLICATION_DOC: "application/msword",
        APPLICATION_DOCX: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        APPLICATION_XLS: "application/vnd.ms-excel",
        APPLICATION_XLSX: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        TEXT_CSV: "text/csv",
        APPLICATION_PPT: "application/vnd.ms-powerpoint",
        APPLICATION_PPTX: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        AUDIO_WAV: "audio/wav",
        AUDIO_X_WAV: "audio/x-wav", //Firefox
        AUDIO_VND_WAVE: "audio/vnd.wave", //IE
        INTERACTIVE_MESSAGE: "application/vnd.amazonaws.connect.message.interactive"
    }
};

ContentType.ATTACHMENT_CONTENT_TYPE = {
    TXT: ContentType.MESSAGE_CONTENT_TYPE.TEXT_PLAIN,
    PDF: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PDF,
    JPG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_JPG,
    PNG: ContentType.MESSAGE_CONTENT_TYPE.IMAGE_PNG,
    DOC: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOC,
    DOCX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_DOCX,
    XLS: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLS,
    XLSX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_XLSX,
    CSV: ContentType.MESSAGE_CONTENT_TYPE.TEXT_CSV,
    PPT: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPT,
    PPTX: ContentType.MESSAGE_CONTENT_TYPE.APPLICATION_PPTX,
    WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_WAV,
    X_WAV: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_X_WAV,
    VND_WAVE: ContentType.MESSAGE_CONTENT_TYPE.AUDIO_VND_WAVE
};

//OpenXML content types do not show up in custom files list unless file extension is explicitly provided
export const ATTACHMENT_ACCEPT_CONTENT_TYPES = [
    ...Object.values(ContentType.ATTACHMENT_CONTENT_TYPE),
    ".docx",
    ".xlsx",
    ".pptx"
];

export const InteractiveMessageType = {
    LIST_PICKER: "ListPicker",
    TIME_PICKER: "TimePicker",
    PANEL: "Panel"
};

export const Status = {
    Sending: "Sending",
    SendSuccess: "SendSuccess",
    SendFailed: "SendFailed",
    Read: "Read"
    // Delivered?
};

export const AttachmentStatus = {
    APPROVED: "APPROVED",
    REJECTED: "REJECTED"
};

export const Direction = {
    Outgoing: "Outgoing",
    Incoming: "Incoming"
};

export const AttachmentErrorType = {
    AccessDeniedException: "AccessDeniedException",
    InternalServerException: "InternalServerException",
    ThrottlingException: "ThrottlingException",
    ValidationException: "ValidationException",
    ServiceQuotaExceededException: "ServiceQuotaExceededException",
    ResourceNotFoundException: "ResourceNotFoundException",
    ConflictException: "ConflictException"
};
export class TransportDetails {
    constructor(input) {
        var args = input || {};
        this.direction = args.direction;
        this.readTime = args.readTime;
        this.status = args.status;
        this.sentTime = args.sentTime;
    }
}

export class ItemDetails {
    constructor(input) {
        var args = input || {};
        this.id = args.id;
        this.type = args.type;
        this.content = args.content;
        this.displayName = args.displayName;
        this.participantId = args.participantId;
        this.participantRole = args.participantRole;
        this.transportDetails = args.transportDetails;
    }
}
