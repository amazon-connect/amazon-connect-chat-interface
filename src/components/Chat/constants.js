// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

// eslint-disable-next-line
export const URL_MATCH_REGEX = /(\b(https?|ftp|file):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/i;

export const CHAT_STATUS_TYPES = {
  CONNECTED: "connected",
  MISSED: "missed",
  ACW: "acw",
  INCOMING: "incoming",
  DISCONNECTED: "disconnected",
  CONNECTING: "connecting",
  ENDED: "ended",
  ACCEPTED: "accepted",
  ERROR: "error",
};

export const CHAT_FEATURE_TYPES = {
  ATTACHMENTS: "ATTACHMENTS",
};

export const FAIL_AFTER_NO_RETURN_SECONDS = 10;

export const INTERACTIVE_MESSAGE = {
  ACTIONS: {
    SHOW_MORE: "SHOW_MORE",
    PREVIOUS_OPTIONS: 'PREVIOUS_OPTIONS'
  },
  VERSION: "1.0"
}
