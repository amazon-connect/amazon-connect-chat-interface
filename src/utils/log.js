/* Add your own logger function here
  var customizedLogger = {
    debug: (data) => {// customize logger function here},
    info: (data) => {// customize logger function here},
    error: (data) => {// customize logger function here}
  }
  "connect" is a global object that exported from src/components/Chat/amazon-connect-chat.js
  There are four levels available from connect.LogLevel - DEBUG, INFO, ERROR.
*/
export const config = {
    loggerConfig: {
        // Add your logger here
        // customizedLogger: customizedLogger,
        level: connect.LogLevel.INFO,
        useDefaultLogger: true
    },
};
