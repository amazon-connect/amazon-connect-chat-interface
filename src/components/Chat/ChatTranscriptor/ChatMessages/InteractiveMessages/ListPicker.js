// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React, { useState } from "react";
import styled from "styled-components";
import PT from "prop-types";
import { Button } from "connect-core";
import {
  ReactiveImage,
  TextSection,
  ResponsesSection,
  HeaderText,
  Title,
  Subtitle,
  PickerOptionTitle,
  PickerElementLinkOption
} from "../InteractiveMessage";
import { createInteractiveMessagePayload, truncateStrFromCharLimit } from "../../../../../utils/helper";
import { InteractiveMessageSelectionType, InteractiveMessageType } from "../../../datamodel/Model";
import { RichMessageRenderer } from "../../../RichMessageComponents";

//#region Styled Components
// have to calculate max height to maintain 16:9 aspect ratio
const ImageContainer = styled.div`
  max-height: calc(95vw * (9 / 16));
  overflow: hidden;
  display: ${(props) => (props.showImage ? "flex" : "none")};

  ${props => props.isCarouselElem ? `
    img {
      float: left;
      height: 10rem;
      object-fit: cover;
    }
  ` : ""}
`;

// Align-right when using subtitle and/or image - otherwise center the text
// Action buttons(SHOW_MORE/PREVIOUS_OPTIONS) always align to center
const ListElementButton = styled(Button)`
  display: flex;
  justify-content: ${(props) => (props.isFloatLeft ? "flex-start" : "center")};
  width: 100%;
  max-width: none;
  border: ${({ theme }) => theme.globals.baseBorder};
  white-space: pre-line;

  &:hover {
    background: ${({ theme }) => theme.color.primary};
  }

  /* Last child could be <a/>, only round last child */
  &:last-child {
    border-bottom-left-radius: ${({ theme }) => theme.spacing.mini};
    border-bottom-right-radius: ${({ theme }) => theme.spacing.mini};
    margin-bottom: 0;
  }
`;

const ElementImageContainer = styled.div`
  width: ${(props) => props.theme.spacing.xxlarge};
  height: ${(props) => props.theme.spacing.xxlarge};
  border: ${({ theme }) => theme.globals.baseBorder};
  border-radius: ${({ theme }) => theme.spacing.mini};
  flex-shrink: 0;
  margin-right: ${({ theme }) => theme.spacing.small};
  display: ${(props) => (props.showImage ? "inline-flex" : "none")};
`;
//#endregion Styled Components

function ListPickerElement({
  element,
  onClick,
  showImage,
  onImageLoad,
  index,
}) {
  const { title: inputTitle, subtitle: inputSubtitle, imageData, imageDescription, type, url } = element;
  const title = truncateStrFromCharLimit(inputTitle, InteractiveMessageType.LIST_PICKER, "elementTitleCharLimit");
  const subtitle = truncateStrFromCharLimit(inputSubtitle, InteractiveMessageType.LIST_PICKER, "elementSubtitleCharLimit");

  if (type === InteractiveMessageSelectionType.HYPERLINK && url) {
    return (
      <PickerElementLinkOption
        {...element}
        testId={"listElement" + index}
        title={title}
      />
    )
  }

  return (
    <ListElementButton
      value={title}
      isFloatLeft={(showImage && imageData) || subtitle}
      onClick={onClick}
      data-testid={"listElement" + index}
    >
      {imageData && (
        <ElementImageContainer showImage={showImage}>
          <ReactiveImage
            imageSrc={imageData}
            imageDescription={imageDescription}
            onImageLoad={onImageLoad}
          />
        </ElementImageContainer>
      )}
      {subtitle ? (
        <TextSection>
            <RichMessageRenderer content={title} styledWrapper={Title} />
            <RichMessageRenderer content={subtitle} styledWrapper={Subtitle} />
        </TextSection>
      ) : (
        <PickerOptionTitle>{title}</PickerOptionTitle>
      )}
    </ListElementButton>
  );
}

ListPicker.propTypes = {
  content: PT.object.isRequired,
  addMessage: PT.func.isRequired,
  isCarouselElem: PT.bool,
  templateIdentifier: PT.string
};

export default function ListPicker({
  content,
  addMessage,
  templateType,
  isCarouselElem,
  templateIdentifier
}) {
  // assumptions: version 1, image data is URL. Guarenteed title exists, at least 1 element.
  const {
    title: inputTitle,
    subtitle: inputSubtitle,
    elements: inputElements,
    imageData,
    imageDescription,
    preIndex,
    nextIndex,
    listId,
    referenceId,
  } = content;

  // Frontend field validations
  const title = truncateStrFromCharLimit(inputTitle, InteractiveMessageType.LIST_PICKER, "titleCharLimit");
  const subtitle = truncateStrFromCharLimit(inputSubtitle, InteractiveMessageType.LIST_PICKER, "subtitleCharLimit");

  const [imageLoaded, setImageLoaded] = useState(false);
  const [elementImagesLoadedCount, setElementImagesLoadedCount] = useState(0);

  function onImageLoad() {
    setImageLoaded(true);
  }

  function onElementImageLoad() {
    setElementImagesLoadedCount(elementImagesLoadedCount + 1);
  }

  function onItemClick(index) {
    const selectedElement = inputElements[index];
    const payload = createInteractiveMessagePayload(
      selectedElement,
      preIndex,
      nextIndex,
      listId,
      templateType,
      referenceId,
      isCarouselElem,
      title,
      templateIdentifier
    );
    addMessage(payload);
  }

  const showElementImages = () => {
    const actionBtnCount = inputElements.filter((ele) => {
      return ele.actionDetail;
    }).length;
    return elementImagesLoadedCount === inputElements.length - actionBtnCount;
  };

  return (
    <>
      {imageData && (
        <ImageContainer showImage={imageLoaded} isCarouselElem={isCarouselElem}>
          <ReactiveImage
            imageSrc={imageData}
            imageDescription={imageDescription}
            onImageLoad={onImageLoad}
          />
        </ImageContainer>
      )}
      <HeaderText title={title} subtitle={subtitle} />
      <ResponsesSection isCarouselElem={isCarouselElem}>
        <div>
          {inputElements.map((listPickerElement, index) => (
            <ListPickerElement
              key={"element-" + index}
              element={listPickerElement}
              onClick={() => {
                onItemClick(index);
              }}
              showImage={showElementImages()}
              onImageLoad={onElementImageLoad}
              index={index}
            />
          ))}
        </div>
      </ResponsesSection>
    </>
  );
}
