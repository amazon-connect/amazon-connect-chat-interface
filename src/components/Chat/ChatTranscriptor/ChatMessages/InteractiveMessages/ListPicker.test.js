// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { ThemeProvider } from "../../../../../theme";
import ListPicker from "./ListPicker";
import { InteractiveMessageType } from "../../../datamodel/Model";
import { INTERACTIVE_MESSAGE } from "../../../constants";
import "@testing-library/jest-dom";
import * as helpers from '../../../../../utils/helper';

const LIST_PICKER_CONSTRAINTS = helpers.INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType.LIST_PICKER];

const mockListPickerContent = {
  title: "ListPickerTitle",
  subtitle: "ListPickerSubtitle",
  elements: [
    {
      title: "ListPickerElementTitle",
      subtitle: "ListPickerElementSubtitle",
      imageData: "ListPickerElementImageData",
      imageDescription: "ListPickerElementImageDescription",
    },
    {
      title: "AnotherListPickerElementTitle",
      subtitle: "AnotherListPickerElementSubtitle",
      imageData: "AnotherListPickerElementImageData",
      imageDescription: "AnotherListPickerElementImageDescription",
    },
  ],
  imageData: "ListPickerImageData",
  imageDescription: "ListPickerImageDescription",
};

export const mockListPickerData = JSON.stringify({
  templateType: InteractiveMessageType.LIST_PICKER,
  isCustomInteractive: "true",
  version: "1.0",
  data: {
    content: mockListPickerContent,
  },
});

describe("<ListPicker />", () => {
  let mockProps;

  function renderElement(props) {
    return render(
      <ThemeProvider>
        <ListPicker {...props} />
      </ThemeProvider>
    );
  }

  beforeEach(() => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = { content: mockListPickerContent, addMessage: addMessage, templateType: InteractiveMessageType.LIST_PICKER };
  });

  it("style should match the snapshot", () => {
    const mockListPicker = renderElement(mockProps);
    expect(mockListPicker).toMatchSnapshot();
  });

  it("should be able to use ListPicker", () => {
    renderElement(mockProps);

    expect(screen.getByText("ListPickerTitle")).toBeDefined();
    expect(screen.getByText("ListPickerSubtitle")).toBeDefined();
    expect(screen.getByAltText("ListPickerImageDescription")).toBeDefined();

    expect(screen.getByText("ListPickerElementTitle")).toBeDefined();
    expect(screen.getByText("AnotherListPickerElementTitle")).toBeDefined();
    expect(
      screen.getByAltText("ListPickerElementImageDescription")
    ).toBeDefined();
    expect(
      screen.getByAltText("AnotherListPickerElementImageDescription")
    ).toBeDefined();

    fireEvent.click(screen.getByText("ListPickerElementTitle"));
    expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
    expect(mockProps.addMessage).toHaveBeenCalledWith({
      text: "ListPickerElementTitle",
    });
  });

  it("Should create interactive selection message payload", () => {
    renderElement(mockProps);

    const createMsgPayloadSpy = jest.spyOn(helpers, 'createInteractiveMessagePayload');
    fireEvent.click(screen.getByText("ListPickerElementTitle"));
    expect(createMsgPayloadSpy).toHaveBeenCalledWith(mockListPickerContent.elements[0], undefined, undefined, undefined, InteractiveMessageType.LIST_PICKER, undefined, undefined, "ListPickerTitle", undefined);

    createMsgPayloadSpy.mockRestore();
  });

  it.each([
      ["title", LIST_PICKER_CONSTRAINTS.titleCharLimit],
      ["subtitle", LIST_PICKER_CONSTRAINTS.subtitleCharLimit]
  ])("should truncate a %s over the character limit", (fieldKey, charLimit) => {
      const longTitle = "LongTitle".repeat(150);
      const truncatedTitle = `${longTitle.substring(0, charLimit)}...`;

      const listPickerLongTitle = {
          ...mockListPickerContent,
          [fieldKey]: longTitle,
      };
      renderElement({ ...mockProps, content: listPickerLongTitle });
      expect(() => screen.getByText(longTitle)).toThrow(
          "Unable to find an element"
      );
      expect(() => screen.getByText(truncatedTitle)).not.toThrow();
  });

  it.each([
      ["title", LIST_PICKER_CONSTRAINTS.elementTitleCharLimit],
      ["subtitle", LIST_PICKER_CONSTRAINTS.elementSubtitleCharLimit],
  ])("should truncate an element %s over the character limit", (fieldKey, charLimit) => {
      const longElementTitle = "LongElementTitle".repeat(100);
      const truncatedElementTitle = `${longElementTitle.substring(0, charLimit)}...`;

      const listPickerPickerLongTitle = {
          ...mockListPickerContent,
          elements: [...mockListPickerContent.elements, {
            ...mockListPickerContent.elements[0],
            [fieldKey]: longElementTitle
          }]
      };
      renderElement({ ...mockProps, content: listPickerPickerLongTitle });
      expect(() => screen.getByText(longElementTitle)).toThrow(
          "Unable to find an element"
      );
      expect(() => screen.getByText(truncatedElementTitle)).not.toThrow();

      // Should still send entire title to backend, instead of returning 'my selectio...'
      fireEvent.click(screen.getByText(truncatedElementTitle));
      expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
      expect(mockProps.addMessage).toHaveBeenCalledWith({
        text: listPickerPickerLongTitle.elements[listPickerPickerLongTitle.elements.length - 1]["title"],
      });
  });
});

describe("ListPicker XSS Mitigation", () => {
  let mockProps;
  const mockListPickerXSSContent = {
    title: 'Title with script <script>alert("XSS attack!");</script>',
    subtitle: '<img src="x" onerror="alert(\'XSS attack!\');">',
    elements: [
      {
        title: '<a href="javascript:alert(\'XSS attack!\')">Click me</a>',
        subtitle: "Subtitle with script! <script>alert(\"XSS attack!\");</script>",
      },
      {
        title: '<input type="text" value="XSS attack!" onfocus="alert(\'XSS attack!\');">',
        subtitle: "Subtitle with script! <script>alert(\"XSS attack!\");</script>",
      },
      {
        title: '<div data-value="<img src=x onerror=alert(\'XSS attack!\')>"></div>',
        subtitle: "Subtitle with script! <script>alert(\"XSS attack!\");</script>",
      },
      {
        title: '<div style="background-image: url(\'javascript:alert(\'XSS attack!\');\')"></div>',
        subtitle: "Subtitle with script! <script>alert(\"XSS attack!\");</script>",
      }
    ],
  };
  beforeEach(() => {
    mockProps = {
      content: mockListPickerXSSContent,
      addMessage: jest.fn(),
      templateType: InteractiveMessageType.LIST_PICKER,
    };
  });
  function renderElement(props) {
    return render(
      <ThemeProvider>
        <ListPicker {...props} />
      </ThemeProvider>
    );
  }

  it("should use DOMPurify to mitigate malicious XSS input", () => {
    renderElement(mockProps);
    expect(screen.getByText("Title with script")).toBeDefined();
    expect(screen.getByText("<img src=\"x\">", { exact: false })).toBeDefined();
    expect(screen.getByText("<a>Click me</a>", { exact: false })).toBeDefined();
    expect(screen.getByText("<input value=\"XSS attack!\" type=\"text\">", { exact: false })).toBeDefined();
    expect(screen.getByText("<div data-value=\"<img src=x onerror=alert('XSS attack!')>\"></div>", { exact: false })).toBeDefined();
    expect(screen.getByText("<div style=\"background-image: url('javascript:alert('XSS attack!');')\"></div>", { exact: false })).toBeDefined();
  });
});


describe("ListPicker with action buttons test", () => {
  let mockProps;
  const mockListPickerContent = {
    title: "ListPickerTitle",
    subtitle: "ListPickerSubtitle",
    preIndex: "preIndex",
    nextIndex: "nextIndex",
    listId: "listId",
    referenceId: "referenceId",
    elements: [
      {
        title: "Previous options",
        actionDetail: INTERACTIVE_MESSAGE.ACTIONS.PREVIOUS_OPTIONS,
      },
      {
        title: "AnotherListPickerElementTitle1",
        subtitle: "AnotherListPickerElementSubtitle1",
        imageData: "AnotherListPickerElementImageData1",
        imageDescription: "AnotherListPickerElementImageDescription1",
      },
      {
        title: "AnotherListPickerElementTitle2",
        subtitle: "AnotherListPickerElementSubtitle2",
        imageData: "AnotherListPickerElementImageData2",
        imageDescription: "AnotherListPickerElementImageDescription2",
      },
      {
        title: "Show more",
        actionDetail: INTERACTIVE_MESSAGE.ACTIONS.SHOW_MORE,
      },
    ],
    imageData: "ListPickerImageData",
    imageDescription: "ListPickerImageDescription",
  };
  beforeEach(() => {
    const addMessage = jest.fn().mockResolvedValue(undefined);
    mockProps = {
      content: mockListPickerContent,
      addMessage: addMessage,
      templateType: InteractiveMessageType.LIST_PICKER,
    };
  });
  function renderElement(props) {
    return render(
      <ThemeProvider>
        <ListPicker {...props} />
      </ThemeProvider>
    );
  }
  it("should be able to use action buttons in list picker", () => {
    renderElement(mockProps);
    expect(screen.getByText("Previous options")).toBeDefined();
    expect(screen.getByText("Show more")).toBeDefined();
    expect(screen.getByTestId("listElement0")).toHaveStyle(
      "justify-content: center"
    );
    expect(screen.getByTestId("listElement1")).toHaveStyle(
      "justify-content: flex-start"
    );
    expect(screen.getByTestId("listElement3")).toHaveStyle(
      "justify-content: center"
    );
    fireEvent.click(screen.getByText("Previous options"));
    expect(mockProps.addMessage).toBeCalledWith({
      text: '{"version":"1.0","data":{"actionName":"PREVIOUS_OPTIONS","preIndex":"preIndex","nextIndex":"nextIndex","listId":"listId","templateType":"ListPicker","referenceId":"referenceId"},"action":"PREVIOUS_OPTIONS"}',
      type: "application/vnd.amazonaws.connect.message.interactive.response",
    });
    fireEvent.click(screen.getByText("Show more"));
    expect(mockProps.addMessage).toBeCalledWith({
      text: '{"version":"1.0","data":{"actionName":"SHOW_MORE","preIndex":"preIndex","nextIndex":"nextIndex","listId":"listId","templateType":"ListPicker","referenceId":"referenceId"},"action":"SHOW_MORE"}',
      type: "application/vnd.amazonaws.connect.message.interactive.response",
    });
  });
});
