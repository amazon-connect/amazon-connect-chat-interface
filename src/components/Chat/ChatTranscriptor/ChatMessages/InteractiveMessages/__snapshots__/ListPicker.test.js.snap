// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListPicker /> style should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c3 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  color: inherit;
  margin: 0;
}

.c3:last-child {
  margin: 0;
}

.c7 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c7 > img,
.c7 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c7:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c7:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c7:active {
  border-color: #fff #fff #fff;
}

.c7:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c7:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c4 {
  font-weight: bold !important;
}

.c5 {
  color: #555555;
}

.c1 {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c2 {
  padding: 14px;
  text-align: left;
}

.c6 {
  position: relative;
  white-space: pre-line;
  border-radius: 6px;
}

.c0 {
  max-height: calc(95vw * (9 / 16));
  overflow: hidden;
  display: none;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  width: 100%;
  max-width: none;
  border: 1px solid #DDDDDD;
  white-space: pre-line;
}

.c8:hover {
  background: #06adcb;
}

.c8:last-child {
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: 0;
}

.c9 {
  width: 44px;
  height: 44px;
  border: 1px solid #DDDDDD;
  border-radius: 6px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 10px;
  display: none;
}

<body>
    <div>
      <div
        class="c0"
      >
        <img
          alt="ListPickerImageDescription"
          class="c1"
          src="ListPickerImageData"
        />
      </div>
      <div
        class="c2"
      >
        <p
          class="c3 c4"
          style="margin: 0px;"
        >
          ListPickerTitle
        </p>
        <p
          class="c3 c5"
          style="margin: 0px;"
        >
          ListPickerSubtitle
        </p>
      </div>
      <div
        class="c6"
      >
        <div>
          <button
            class="c7 c8"
            data-selected="false"
            data-testid="listElement0"
            type="default"
            value="ListPickerElementTitle"
          >
            <div
              class="c9"
            >
              <img
                alt="ListPickerElementImageDescription"
                class="c1"
                src="ListPickerElementImageData"
              />
            </div>
            <div
              class="c2"
            >
              <p
                class="c3 c4"
                style="margin: 0px;"
              >
                ListPickerElementTitle
              </p>
              <p
                class="c3 c5"
                style="margin: 0px;"
              >
                ListPickerElementSubtitle
              </p>
            </div>
          </button>
          <button
            class="c7 c8"
            data-selected="false"
            data-testid="listElement1"
            type="default"
            value="AnotherListPickerElementTitle"
          >
            <div
              class="c9"
            >
              <img
                alt="AnotherListPickerElementImageDescription"
                class="c1"
                src="AnotherListPickerElementImageData"
              />
            </div>
            <div
              class="c2"
            >
              <p
                class="c3 c4"
                style="margin: 0px;"
              >
                AnotherListPickerElementTitle
              </p>
              <p
                class="c3 c5"
                style="margin: 0px;"
              >
                AnotherListPickerElementSubtitle
              </p>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jbKcbu jNUAuY"
    >
      <img
        alt="ListPickerImageDescription"
        class="sc-ksYbfQ dfzqNG"
        src="ListPickerImageData"
      />
    </div>
    <div
      class="sc-hmzhuo fLEfL"
    >
      <p
        class="sc-bZQynM fraGhu sc-kgAjT hWSRJP"
        style="margin: 0px;"
      >
        ListPickerTitle
      </p>
      <p
        class="sc-bZQynM fraGhu sc-cJSrbW bjtoOs"
        style="margin: 0px;"
      >
        ListPickerSubtitle
      </p>
    </div>
    <div
      class="sc-frDJqD cZGRLS"
    >
      <div>
        <button
          class="sc-gzVnrw fgSGc sc-dNLxif ezVYkD"
          data-selected="false"
          data-testid="listElement0"
          type="default"
          value="ListPickerElementTitle"
        >
          <div
            class="sc-jqCOkK exNHjp"
          >
            <img
              alt="ListPickerElementImageDescription"
              class="sc-ksYbfQ dfzqNG"
              src="ListPickerElementImageData"
            />
          </div>
          <div
            class="sc-hmzhuo fLEfL"
          >
            <p
              class="sc-bZQynM fraGhu sc-kgAjT hWSRJP"
              style="margin: 0px;"
            >
              ListPickerElementTitle
            </p>
            <p
              class="sc-bZQynM fraGhu sc-cJSrbW bjtoOs"
              style="margin: 0px;"
            >
              ListPickerElementSubtitle
            </p>
          </div>
        </button>
        <button
          class="sc-gzVnrw fgSGc sc-dNLxif ezVYkD"
          data-selected="false"
          data-testid="listElement1"
          type="default"
          value="AnotherListPickerElementTitle"
        >
          <div
            class="sc-jqCOkK exNHjp"
          >
            <img
              alt="AnotherListPickerElementImageDescription"
              class="sc-ksYbfQ dfzqNG"
              src="AnotherListPickerElementImageData"
            />
          </div>
          <div
            class="sc-hmzhuo fLEfL"
          >
            <p
              class="sc-bZQynM fraGhu sc-kgAjT hWSRJP"
              style="margin: 0px;"
            >
              AnotherListPickerElementTitle
            </p>
            <p
              class="sc-bZQynM fraGhu sc-cJSrbW bjtoOs"
              style="margin: 0px;"
            >
              AnotherListPickerElementSubtitle
            </p>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
